plugins/woocommerce/src/Admin/Features/ProductBlockEditor/BlockTemplateUtils.php

```diff
@@ -133,15 +133,65 @@ private function create_new_block_template( $template_file, $template_type, $tem
         }
         $blocks            = parse_blocks( $template->content );
         $template->content = traverse_and_serialize_blocks( $blocks, $before_block_visitor, $after_block_visitor );
 
 		return $template;
 	}
 
+    /**
+     * Returns the markup for blocks hooked to the given anchor block in a specific relative position.
+     *
+     * @since 6.5.0
+     * @access private
+     *
+     * @param array                   $parsed_anchor_block The anchor block, in parsed block array format.
+     * @param string                  $relative_position   The relative position of the hooked blocks.
+     *                                                     Can be one of 'before', 'after', 'first_child', or 'last_child'.
+     * @param array                   $hooked_blocks       An array of hooked block types, grouped by anchor block and relative position.
+     * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
+     * @return string
+     */
+    public function insert_blocks( &$parsed_anchor_block, $relative_position, $hooked_blocks, $context ) {
+        /**
+         * Filters the list of inserted blocks for a given anchor block type and relative position.
+         *
+         * @param array[]                         $inserted_blocks     The list of inserted blocks.
+         * @param string                          $parsed_anchor_block The parsed anchor block.
+         * @param string                          $relative_position   The relative position of the hooked blocks.
+         *                                                             Can be one of 'before', 'after', 'first_child', or 'last_child'.
+         * @param WP_Block_Template|WP_Post|array $context             The block template, template part, `wp_navigation` post type,
+         *                                                             or pattern that the anchor block belongs to.
+         */
+        $inserted_blocks = apply_filters( "inserted_blocks", array(), $parsed_anchor_block, $relative_position, $context );
+
+        $markup = '';
+        foreach ( $inserted_blocks as $inserted_block ) {
+            $parsed_inserted_block = array(
+                'blockName'    => $inserted_block['blockName'] ?? '',
+                'attrs'        => $inserted_block['attrs'] ?? array(),
+                'innerBlocks'  => $inserted_block['innerBlocks'] ?? array(),
+                'innerContent' => $inserted_block['innerContent'] ?? array(),
+            );
+
+            // Parse these first so markup is generated from all inner blocks.
+            $first_chunk = $this->insert_blocks( $parsed_inserted_block, 'first_child', $hooked_blocks, $context );
+            $last_chunk  = $this->insert_blocks( $parsed_inserted_block, 'last_child', $hooked_blocks, $context );
+            array_unshift( $parsed_inserted_block['innerContent'], $first_chunk );
```

Comment:
This needs further investigation to make sure this won't break things.  I suspect this could be problematic depending on existing chunks in `innerContent`.

Reply:
I think you need to prepend/append the hooked block to `innerBlocks` rather than `innerContent`:

```suggestion
            array_unshift( $parsed_inserted_block['innerBlocks'], $first_chunk );
```

Additionally, you also need to insert a `null` in the corresponding position in `innerContent`, so it should actually read

```suggestion
            array_unshift( $parsed_inserted_block['innerBlocks'], $first_chunk );
            array_unshift( $parsed_inserted_block['innerContent'], null );
```

The way it works is that `innerContent` can have items that are either HTML strings, or `null` to mark a position where an item from `innerBlocks` needs to be interpolated. E.g.

```php
$group_block_with_some_inner_block = array(
	'blockName' => 'core/group',
	'attrs'     => array(),
	'innerBlocks' => array( $some_inner_block ),
	'innerContent' => array(
		'<div class="wp-block-group">',
		null,
		'</div>'
	),
);
```

Reply:
This is still not the full picture BTW if you want to account for some subtleties: As the Group block example demonstrates, you might not actually want to put the `first_child` in the very first position (or the `last_child` in the very last one), as you might want to account for the wrapping HTML:

```
$group_block_with_some_inner_block = array(
	'blockName' => 'core/group',
	'attrs'     => array(),
	'innerBlocks' => array( $hooked_first_child, $some_inner_block ),
	'innerContent' => array(
		null, // $hooked_first_child is going to be rendered outside the wrapping div :/
		'<div class="wp-block-group">',
		null,
		'</div>'
	),
);
```

Instead, you'll likely want to find the first non-HTML location in `innerContent`. For the original Block Hooks implementation (in GB's WP 6.4 compat shim), we ended up with the following logic: https://github.com/WordPress/gutenberg/blob/98552b2486b52e4dd3c90775c2a029b95e64253c/lib/compat/wordpress-6.4/block-hooks.php#L157-L182